### PR TITLE
CI tests.yml: Deploy the retired Docker image

### PR DESCRIPTION
**Pull request by sjackman**
Merge this PR after merging PR https://github.com/Homebrew/homebrew-core/pull/108590

- https://github.com/Homebrew/homebrew-core/pull/108590

-----

- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- <https://github.com/Homebrew/brew/issues/13619>

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -188,6 +188,16 @@ jobs:
...
 
+      - name: Deploy the retired Docker image to GitHub Packages and Docker Hub
+        if: github.ref == 'refs/heads/master'
+        run: |
+          printf 'FROM alpine\nCMD echo "Error: homebrew/ubuntu16.04:master is retired. Use homebrew/brew or homebrew/ubuntu16.04 or homebrew/ubuntu22.04:master"; exit 1\n' |
+            docker build --tag=retired -
+          docker tag retired "ghcr.io/homebrew/ubuntu16.04:master"
+          docker push "ghcr.io/homebrew/ubuntu16.04:master"
+          docker tag retired "homebrew/ubuntu16.04:master"
+          docker push "homebrew/ubuntu16.04:master"
```
Comment (carlocab) [inline .github/workflows/tests.yml:199]: Seems weird to do this every time when it's something that only needs to be done once.

Comment (sjackman) [reply .github/workflows/tests.yml:199]: We can revert this PR immediately after it's merged.

**Review by carlocab — APPROVED** [review]
Seems better suited for its own workflow that is done with `workflow_dispatch`, but I'm not strongly opposed to this.

**Comment by Bo98** [comment]
Tested this out. The message is never displayed in GitHub Actions: https://github.com/Bo98/brew-image-retirement-test/runs/8198518314?check_suite_focus=true

I think what we would have to do is keep a Ubuntu 16.04 base (for bash etc.) and have a stub `brew` script that prints the message. (This can be addition to the `CMD`/`ENTRYPOINT`.)

If we'd like a deprecation period, which I would encourage, I have played with that a bit. Looks like this: https://github.com/Bo98/brew-image-retirement-test/runs/8199992071?check_suite_focus=true (ignore the permission error). Could be tweaked to be less spammy but seems to work. Also appears in the summary: https://github.com/Bo98/brew-image-retirement-test/actions/runs/2997682091, which is also where GitHub puts their own deprecation messages. Code: https://github.com/Bo98/brew/commit/e36427282c425220b9d84e760a289ad6482ed172 + a `.docker-deprecate` file planted in the Docker image.

**Comment by carlocab** [comment]
> Also appears in the summary: https://github.com/Bo98/brew-image-retirement-test/actions/runs/2997682091, which is also where GitHub puts their own deprecation messages.

I like this.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (MikeMcQuaid) [reply .github/workflows/tests.yml:199]: How about guarding whatever we go with to the 3.6.0 tag? Would like to have the code kept around so we have it for the next Ubuntu LTS upgrade. Alternatively, could just add it to the docs.

**Comment by Bo98** [comment]
#13819 shows what we can do for a crossover deprecation period.

What we do for retirement will be slightly different. I'll open a second PR for that.

**Pull request closed without merging**